src/smartcontracts/nativeSerializer.spec.ts

```diff
@@ -45,6 +46,87 @@ describe("test native serializer", () => {
         assert.deepEqual(typedValues[6].valueOf(), null);
     });
 
+    it("should perform type inference (regular variadic arguments)", async () => {
+        const endpointModifiers = new EndpointModifiers("", []);
+        const inputParameters = [new EndpointParameterDefinition("", "", new VariadicType(new U32Type(), false))];
+        const endpoint = new EndpointDefinition("foo", inputParameters, [], endpointModifiers);
+        const typedValues = NativeSerializer.nativeToTypedValues([8, 9, 10], endpoint);
+
+        assert.deepEqual(typedValues[0].getType(), new VariadicType(new U32Type(), false));
+        assert.deepEqual(typedValues[0].valueOf(), [new BigNumber(8), new BigNumber(9), new BigNumber(10)]);
+    });
+
+    it("should perform type inference (counted-variadic arguments)", async () => {
+        const endpointModifiers = new EndpointModifiers("", []);
+        const inputParameters = [
+            new EndpointParameterDefinition("", "", new VariadicType(new U32Type(), true)),
+            new EndpointParameterDefinition("", "", new VariadicType(new BytesType(), true))
+        ];
+        const endpoint = new EndpointDefinition("foo", inputParameters, [], endpointModifiers);
+
+        // Implicit counted-variadic (not supported).
+        assert.throws(() => NativeSerializer.nativeToTypedValues([8, 9, 10, "a", "b", "c"], endpoint), ErrInvalidArgument);
+
+        // Explicit, non-empty counted-variadic.
+        let typedValues = NativeSerializer.nativeToTypedValues([
+            VariadicValue.fromItemsCounted(new U32Value(8), new U32Value(9), new U32Value(10)),
+            VariadicValue.fromItemsCounted(BytesValue.fromUTF8("a"), BytesValue.fromUTF8("b"), BytesValue.fromUTF8("c"))
```

Comment:
It's possible to have multiple counted variadic arguments (as users are required to pass them in the explicitly typed form).